### PR TITLE
8324890: C2 SuperWord: refactor out VLoop, make unrolling_analysis static, remove init/reset mechanism

### PR DESCRIPTION
**Pull request by eme64**
Subtask of https://github.com/openjdk/jdk/pull/16620
(The basic goal is to break SuperWord into different modules. This makes the code more maintainable and extensible. And eventually this allows some modules to be reused by other/new vectorizers.)

1. Move out the shared code between `SuperWord::SLP_extract` (where we do vectorization) and `SuperWord::unrolling_analysis`, and move it to a new class `VLoop`. This allows us to decouple `unrolling_analysis` from the SuperWord object, and we can make it static.
2. So far, SuperWord was reused for all loops in a compilation, and then "reset" (with `SuperWord::init`) for every loop. This is a bit of a nasty pattern. I now make a new `VLoop` and a new `SuperWord` object per loop.
3. Since we now make more `SuperWord` objects, we allocate the internal data structures more often. Therefore, I now pre-allocate/reserve sufficient space on initialization.

Side-note about https://github.com/openjdk/jdk/pull/17604 (integrated, no need to read any more):
I would like to remove the use of `SuperWord::is_marked_reduction` from `SuperWord::unrolling_analysis`. For starters: it is not clear what it was ever good for. Second: it requires us to do reduction marking/analysis before `unrolling_analysis`, and hence makes the reduction marking shared between `unrolling_analysis` and vectorization. I could move the reduction marking to `VLoop` now. But the `_loop_reducitons` set would have to be put on an arena, and I would like to avoid creating an arena for the `unrolling_analysis`. Plus, it would just be nicer code, to have reduction analysis together with body analysis, type analysis, etc. and all of them in only in `SLP_extract`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8324890](https://bugs.openjdk.org/browse/JDK-8324890): C2 SuperWord: refactor out VLoop, make unrolling_analysis static, remove init/reset mechanism (**Sub-task** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Roland Westrelin](https://openjdk.org/census#roland) (@rwestrel - **Reviewer**) ⚠️ Review applies to [dcade5f3](https://git.openjdk.org/jdk/pull/17624/files/dcade5f38f37f67f02c7f6810b6a90861098eb71)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17624/head:pull/17624` \
`$ git checkout pull/17624`

Update a local copy of the PR: \
`$ git checkout pull/17624` \
`$ git pull https://git.openjdk.org/jdk.git pull/17624/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17624`

View PR using the GUI difftool: \
`$ git pr show -t 17624`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17624.diff">https://git.openjdk.org/jdk/pull/17624.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17624#issuecomment-1916722213)

### CODE REVIEW

File: src/hotspot/share/opto/superword.cpp [hunk 1]
```diff
@@ -3783,7 +3730,6 @@ void SuperWord::init() {
   _node_info.clear();
   _align_to_ref = nullptr;
   _race_possible = 0;
-  _early_return = false;
```
Comment (eme64) [inline src/hotspot/share/opto/superword.cpp:3786]: remove init!

File: src/hotspot/share/opto/superword.cpp [hunk 2]
```diff
@@ -38,146 +38,51 @@
...
   int max_vector = Matcher::max_vector_size_auto_vectorization(T_BYTE);
 
   // Process the loop, some/all of the stack entries will not be in order, ergo
   // need to preprocess the ignored initial state before we process the loop
-  for (uint i = 0; i < lpt()->_body.size(); i++) {
-    Node* n = lpt()->_body.at(i);
+  for (uint i = 0; i < lpt->_body.size(); i++) {
+    Node* n = lpt->_body.at(i);
     if (n == cl->incr() ||
-      is_marked_reduction(n) ||
```
Comment (eme64) [inline src/hotspot/share/opto/superword.cpp:180]: This is the annoying reduction

File: src/hotspot/share/opto/superword.cpp [hunk 3]
```diff
@@ -38,146 +38,51 @@
...
-  _mem_slice_head(arena(), 8,  0, nullptr),                 // memory slice heads
-  _mem_slice_tail(arena(), 8,  0, nullptr),                 // memory slice tails
-  _node_info(arena(), 8,  0, SWNodeInfo::initial),          // info needed per node
-  _clone_map(phase->C->clone_map()),                        // map of nodes created in cloning
+SuperWord::SuperWord(Arena* arena, const VLoop &vloop) :
+  _vloop(vloop),
+  _arena(arena),
+  _packset(arena, 8,  0, nullptr),                          // packs for the current block
+  _bb_idx(arena, vloop.estimated_node_count(), 0, 0),       // node idx to index in bb
+  _block(arena, 8,  0, nullptr),                            // nodes in current block
```
Comment (eme64) [inline src/hotspot/share/opto/superword.cpp:46]: reserve space!

File: src/hotspot/share/opto/vectorization.cpp [hunk 4]
```diff
@@ -31,6 +31,89 @@
...
+
+#ifndef PRODUCT
+  if (is_trace_precondition()) {
+    tty->print_cr("VLoop::check_precondition: failed: %s", return_state);
+  }
+#endif
+  return false; // failure
+}
+
+const char* VLoop::check_preconditions_helper() {
```
Comment (eme64) [inline src/hotspot/share/opto/vectorization.cpp:57]: Note: replaces most code from the old `SuperWord::transform_loop`

File: src/hotspot/share/opto/superword.cpp [hunk 5]
```diff
@@ -38,146 +38,51 @@
...
-  _early_return(true),                                      // analysis evaluations routine
-  _do_vector_loop(phase->C->do_vector_loop()),              // whether to do vectorization/simd style
+  _do_vector_loop(phase()->C->do_vector_loop()),            // whether to do vectorization/simd style
   _num_work_vecs(0),                                        // amount of vector work we have
   _num_reductions(0)                                        // amount of reduction work we have
 {
 }
 
-//------------------------------transform_loop---------------------------
-bool SuperWord::transform_loop(IdealLoopTree* lpt, bool do_optimization) {
```
Comment (eme64) [inline src/hotspot/share/opto/superword.cpp:75]: Note: code moved to `VLoop::check_preconditions_helper`

File: src/hotspot/share/opto/superword.cpp [hunk 6]
```diff
@@ -38,146 +38,51 @@
...
-      return false;
-    }
-    Node* pre_opaq1 = pre_end->limit();
-    if (pre_opaq1->Opcode() != Op_Opaque1) {
-      return false;
-    }
-    cl->set_pre_loop_end(pre_end);
-  }
-
-  init(); // initialize data structures
```
Comment (eme64) [inline src/hotspot/share/opto/superword.cpp:149]: Note: this is the end of the "preconditions", and we used to set `_early_exit = false` inside `init()`

File: src/hotspot/share/opto/superword.cpp [hunk 7]
```diff
@@ -38,146 +38,51 @@
...
-
-  // Initialize simple data used by reduction marking early.
-  set_lpt(lpt);
-  set_lp(cl);
-  // For now, define one block which is the entire loop body.
-  set_bb(cl);
-
-  if (SuperWordReductions) {
-    mark_reductions();
-  }
```
Comment (eme64) [inline src/hotspot/share/opto/superword.cpp:98]: Note: I now would like to move reduction marking to **after** precondition checking. Hence, I moved it to `SLP_extract`.

File: src/hotspot/share/opto/superword.cpp (hunk 5 above)
Comment (eme64) [reply src/hotspot/share/opto/superword.cpp:75]: Note: all the `do_optimization` parts are not part of preconditions, and hence they are moved to `transform_loop`

File: src/hotspot/share/opto/superword.cpp [hunk 8]
```diff
@@ -484,6 +389,41 @@ void SuperWord::mark_reductions() {
...
+
+#ifndef PRODUCT
+  if (is_trace_superword_any()) {
+    tty->print_cr("\nSuperWord::transform_loop: failed: %s", state);
+  }
+#endif
+  return false;
+}
+
+const char* SuperWord::transform_loop_helper() {
```
Comment (eme64) [inline src/hotspot/share/opto/superword.cpp:414]: TODO remove the helper, we can do that later

File: src/hotspot/share/opto/loopnode.cpp [hunk 9]
```diff
@@ -5962,30 +5948,6 @@ CountedLoopEndNode* CountedLoopNode::find_pre_loop_end() {
...
-#endif
-    return _pre_loop_end;
-  }
-
-  void CountedLoopNode::set_pre_loop_end(CountedLoopEndNode* pre_loop_end) {
-    assert(is_main_loop(), "Only main loop has pre loop");
-    assert(pre_loop_end, "must be valid");
-    _pre_loop_end = pre_loop_end;
-  }
-
```
Comment (eme64) [inline src/hotspot/share/opto/loopnode.cpp:5988]: This should have never been cached in the node itself, but only during autovectorization.

File: src/hotspot/share/opto/loopnode.hpp [hunk 10]
```diff
@@ -231,14 +231,11 @@ class CountedLoopNode : public BaseCountedLoopNode {
   // vector mapped unroll factor here
   int _slp_maximum_unroll_factor;
 
-  // Cached CountedLoopEndNode of pre loop for main loops
-  CountedLoopEndNode* _pre_loop_end;
```
Comment (eme64) [inline src/hotspot/share/opto/loopnode.hpp:235]: Note: this makes the node smaller, and does not cache something that may be invalid later. It was used only during SuperWord. Looks like a bad pattern.

File: src/hotspot/share/opto/superword.cpp [hunk 11]
```diff
@@ -38,146 +38,51 @@
...
-  _iv(nullptr),                                             // induction var
+  _dg(arena),                                               // dependence graph
+  _nlist(arena, vloop.estimated_body_length(), 0, nullptr), // scratch list of nodes
+  _loop_reductions(arena),                                  // reduction nodes in the current loop
   _race_possible(false),                                    // cases where SDMU is true
-  _early_return(true),                                      // analysis evaluations routine
-  _do_vector_loop(phase->C->do_vector_loop()),              // whether to do vectorization/simd style
+  _do_vector_loop(phase()->C->do_vector_loop()),            // whether to do vectorization/simd style
   _num_work_vecs(0),                                        // amount of vector work we have
   _num_reductions(0)                                        // amount of reduction work we have
```
Comment (eme64) [inline src/hotspot/share/opto/superword.cpp:59]: For some data structures I now pre-allocate memory for the maximum size they may ever reach. This is to avoid re-allocation when they grow.
Before this change, we used to only create SuperWord once, and use it for all loops in the compilation. Now that I ripped out the "init" method, and avoid reusing SuperWord this way, we want to make sure we do not re-allocate too much.

File: src/hotspot/share/opto/superword.cpp [hunk 12]
```diff
@@ -38,146 +38,51 @@
...
-    assert(_packset.length() == 0, "packset must be empty");
-    success = SLP_extract();
-  }
-  return success;
-}
+void SuperWord::unrolling_analysis(const VLoop &vloop, int &local_loop_unroll_factor) {
+  IdealLoopTree* lpt    = vloop.lpt();
+  CountedLoopNode* cl   = vloop.cl();
+  Node* cl_exit         = vloop.cl_exit();
+  PhaseIdealLoop* phase = vloop.phase();
```
Comment (eme64) [inline src/hotspot/share/opto/superword.cpp:67]: Made it static, and instead of the SuperWord object, we now only have access to the VLoop object.

File: src/hotspot/share/opto/superword.cpp [hunk 13]
```diff
@@ -519,8 +462,11 @@ void SuperWord::mark_reductions() {
 //    extraction of scalar values from vectors.
 //
 bool SuperWord::SLP_extract() {
-  CountedLoopNode* cl = lpt()->_head->as_CountedLoop();
-  assert(cl->is_main_loop(), "SLP should only work on main loops");
+  assert(cl()->is_main_loop(), "SLP should only work on main loops");
+
+  if (SuperWordReductions) {
+    mark_reductions();
+  }
```
Comment (eme64) [inline src/hotspot/share/opto/superword.cpp:469]: Note: it would be really nice to move reduction marking to all other analysis, like bb construction and velt computation, etc.

File: src/hotspot/share/opto/superword.hpp [hunk 14]
```diff
@@ -319,13 +313,8 @@ class SuperWord : public ResourceObj {
   const GrowableArray<Node*>&      block()   const { return _block; }
   const DepGraph&                  dg()      const { return _dg; }
  private:
-  IdealLoopTree* _lpt;             // Current loop tree node
-  CountedLoopNode* _lp;            // Current CountedLoopNode
   VectorSet      _loop_reductions; // Reduction nodes in the current loop
-  Node*          _bb;              // Current basic block
```
Comment (eme64) [inline src/hotspot/share/opto/superword.hpp:325]: Note: always the same as `cl`

File: src/hotspot/share/opto/vectorization.hpp [hunk 15]
```diff
@@ -48,15 +48,96 @@ class VTrace : public StackObj {
...
+    assert(cl()->is_main_loop(), "only main loop can reference pre-loop");
+    assert(_pre_loop_end != nullptr, "must have found it");
+    return _pre_loop_end;
+  };
+
+  CountedLoopNode* pre_loop_head() const {
+    CountedLoopNode* head = pre_loop_end()->loopnode();
+    assert(head != nullptr, "must find head");
+    return head;
+  };
```
Comment (eme64) [inline src/hotspot/share/opto/vectorization.hpp:101]: Note: before this patch, these two cache-accessors were in the `CounterLoopEndNode`.

File: src/hotspot/share/opto/loopTransform.cpp [hunk 16]
```diff
@@ -1101,12 +1101,9 @@ void IdealLoopTree::policy_unroll_slp_analysis(CountedLoopNode *cl, PhaseIdealLo
...
     if (!cl->was_slp_analyzed()) {
-      SuperWord sw(phase);
-      sw.transform_loop(this, false);
-
-      // If the loop is slp canonical analyze it
-      if (sw.early_return() == false) {
-        sw.unrolling_analysis(_local_loop_unroll_factor);
+      VLoop vloop(this, true);
+      if (vloop.check_preconditions()) {
+        SuperWord::unrolling_analysis(vloop, _local_loop_unroll_factor);
```
Comment (eme64) [inline src/hotspot/share/opto/loopTransform.cpp:1106]: Note: I made `unrolling_analysis` static, and only pass in `vloop`, not all the info in the `SuperWord` object.

File: src/hotspot/share/opto/loopnode.cpp [hunk 17]
```diff
@@ -4863,29 +4862,16 @@ void PhaseIdealLoop::build_and_optimize() {
      C->set_major_progress();
   }
 
-  // Convert scalar to superword operations at the end of all loop opts.
+  // Auto-vectorize main-loop
   if (C->do_superword() && C->has_loops() && !C->major_progress()) {
-    // SuperWord transform
-    SuperWord sw(this);
+    ResourceArea autovectorization_arena;
```
Comment (eme64) [inline src/hotspot/share/opto/loopnode.cpp:4867]: Note: this allows us to free up all the space used by `SuperWord`'s internal data structures between every processed loop. Previously, all internal data structures were on the compile arena.

File: src/hotspot/share/opto/loopopts.cpp [hunk 18]
```diff
@@ -4209,6 +4210,34 @@ bool PhaseIdealLoop::duplicate_loop_backedge(IdealLoopTree *loop, Node_List &old
...
+  if (vloop.check_preconditions()) {
+    // Ensure that all data structures from autovectorization are deallocated later.
+    ResourceMark rm(arena);
+    SuperWord sw(arena, vloop);
+    if (sw.transform_loop()) {
+      return true;
+    }
+  }
+
+  // This counted main-loop either failed preconditions, the analyzer
```
Comment (eme64) [inline src/hotspot/share/opto/loopopts.cpp:4231]:
```suggestion
  // This counted main-loop either failed preconditions,
```

File: src/hotspot/share/opto/loopnode.cpp [hunk 19]
```diff
@@ -4863,29 +4862,16 @@ void PhaseIdealLoop::build_and_optimize() {
      C->set_major_progress();
   }
 
-  // Convert scalar to superword operations at the end of all loop opts.
+  // Auto-vectorize main-loop
   if (C->do_superword() && C->has_loops() && !C->major_progress()) {
```
Comment (vnkozlov) [inline src/hotspot/share/opto/loopnode.cpp:4866]: Should we rename `do_superword()` too?

File: src/hotspot/share/opto/loopopts.cpp [hunk 20]
```diff
@@ -4209,6 +4210,34 @@ bool PhaseIdealLoop::duplicate_loop_backedge(IdealLoopTree *loop, Node_List &old
...
+    }
+  }
+
+  // This counted main-loop either failed preconditions,
+  // or in SuperWord. From now on only unroll the loop.
+  if (cl->has_passed_slp()) {
+    C->set_major_progress();
+    cl->set_notpassed_slp();
+    cl->mark_do_unroll_only();
+  }
```
Comment (vnkozlov) [inline src/hotspot/share/opto/loopopts.cpp:4237]: This is not related to vectorization. Can you move it back and execute based on return value from `autovectorize()` which is not used now?

**Review by vnkozlov — COMMENTED** [review]
In general it looks good. Few comments.

File: src/hotspot/share/opto/loopnode.cpp (hunk 19 above)
Comment (eme64) [reply src/hotspot/share/opto/loopnode.cpp:4866]: @vnkozlov yes, but later, I would say. It simply caries the same values as `UseSuperWord`. I guess eventually we may have different AutoVectorization components, and it may be better to have a unified AutoVectorization option. Is it ok to leave that to the future, and leave it in for now?

**Comment by eme64** [comment]
@vnkozlov I created https://github.com/openjdk/jdk/pull/17683 to add time measurement. Let me try it on a simple benchmark here.

Disabled turbo-boost for benchmark.

**test001,002,003**
`../../../linux-x64/jdk/bin/java -XX:CompileCommand=compileonly,TestAutoVectorizationTime::test0* -Xbatch -XX:+CITime -XX:+UnlockDiagnosticVMOptions -XX:RepeatCompilation=1000 TestAutoVectorizationTime.java`
This patch:
```
Xbatch:
         IdealLoop:            25.508 s
           AutoVectorize:       5.927 s
Xcomp:
         IdealLoop:            12.860 s
           AutoVectorize:       2.736 s
```
TODO

**test101**
`../../../linux-x64/jdk/bin/java -XX:CompileCommand=compileonly,TestAutoVectorizationTime::test1* -Xbatch -XX:+CITime -XX:+UnlockDiagnosticVMOptions -XX:RepeatCompilation=100 TestAutoVectorizationTime.java`
This patch:
```

```
TODO

**test201**
`../../../linux-x64/jdk/bin/java -XX:CompileCommand=compileonly,TestAutoVectorizationTime::test2* -Xbatch -XX:+CITime -XX:+UnlockDiagnosticVMOptions -XX:RepeatCompilation=100 TestAutoVectorizationTime.java`
This patch:
```

```
TODO

File: src/hotspot/share/opto/loopnode.cpp (hunk 19 above)
Comment (vnkozlov) [reply src/hotspot/share/opto/loopnode.cpp:4866]: Yes, we can do it for later.

**Comment by vnkozlov** [comment]
Thank you for running the timing testing.  What about memory fragmentation? Is this code will uses default chunks in Arena (they can be reused) or allocates new chunk (malloc) each time which may lead to fragmentation.

**Comment by eme64** [comment]
@vnkozlov is there a way to measure memory fragmentation? I don't know how to answer that question.
And is there really a difference to how it was done before? Before we just put everything on the `comp_arena`, and never recover the memory until that arena is given up. Now we have a new `autovectorization_arena` for each AutoVectorization pass over all loops. I guess there could be multiple such passes in a single compilation. Before this patch, this means that each such AutoVectorization pass creates its SuperWord object, and allocates memory on the `comp_arena`, and the memory usage of all these passes adds up. With this patch, we at least are able to give up the memory after every pass. Of course this is only helpful if the malloc/free'd chunks can properly be reused.

An alternative: we can put the `autovectorization_arena` at `Compile`. That way, the chunks are kept until the end of compilation, and can be reused between the different AutoVectorization passes/phases.

**Comment by vnkozlov** [comment]
> And I see that the first (init_size) chunk is allocated from a pool. But subsequent chunks (grow) have "non-standard" length, and are malloc/free'd.

Yes, that was my concern.

There are chunks with different sizes: [arena.hpp#L66](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/memory/arena.hpp#L66). Is your allocation sizes > 32*K `Chunk::size` "Default size of an Arena chunk"? `Arena::grow()` uses `MAX2(ARENA_ALIGN(x), (size_t) Chunk::size);`.
Which of SuperWord allocations are big? Can we split them to fit into 32K?

I think, this should not stop you from doing this refactoring. Yes, it will allow return memory back sooner and it is up to OS how it optimize it. I read your offline discussion with Johan. He has interesting suggestion fro growable arrays (use C heap).

**Comment by eme64** [comment]
I think the only really big array is `_bb_idx`, which converts `node->_idx` to `bb_idx`, basically to have smaller idx so that we can then address into more compact arrays. So we need one array that can be as large as the node-count (that can be many K, hence often more than `32K` bytes), and all other arrays can be of the size of the number of nodes in the loop body.

But does using CHeap directly really improve the situation? That just means that we directly malloc, instead of through the Arena, right? This would not make a difference for memory fragmentation, or am I wrong?

But I think it might be a good idea to have at least the large `bb_idx` array be allocated in `Compile`, and we can decide if that is through the `comp_arena` or directly with `CHeap`. That would at least mean that we only have "one unit of fragmentation" per compilation, rather than per SuperWord loop or pass over all loops.

File: src/hotspot/share/opto/superword.cpp [hunk 21]
```diff
@@ -38,144 +38,50 @@
...
+  CountedLoopNode* cl   = vloop.cl();
+  Node* cl_exit         = vloop.cl_exit();
+  PhaseIdealLoop* phase = vloop.phase();
 
-//------------------------------early unrolling analysis------------------------------
-void SuperWord::unrolling_analysis(int &local_loop_unroll_factor) {
   bool is_slp = true;
-  size_t ignored_size = lpt()->_body.size();
+  size_t ignored_size = lpt->_body.size();
   int *ignored_loop_nodes = NEW_RESOURCE_ARRAY(int, ignored_size);
```
Comment (eme64) [inline src/hotspot/share/opto/superword.cpp:71]: TODO: make local, add ResourceMark.

Comment (eme64) [reply src/hotspot/share/opto/superword.cpp:71]: Note: I will do this in a future RFE.

**Comment by eme64** [comment]
@vnkozlov After extensive discussions with @jdksjolen , I now decided to create a `VShareData` class, which has its own arena, which holds the really large array(s), so that they can be shared between the different SuperWord / AutoVectorizations of the loops. This means fragmentation for large arrays is not as low as before this change. All the smaller arrays and small allocations are ok if they are not shared, since they nicely fit in chunks anyway, and therefore we don't have to be worried about fragmentation so much.

File: src/hotspot/share/opto/loopnode.cpp [hunk 22]
```diff
@@ -4863,30 +4863,30 @@ void PhaseIdealLoop::build_and_optimize() {
...
     Compile::TracePhase tp("autoVectorize", &timers[_t_autoVectorize]);
-    // SuperWord transform
-    SuperWord sw(this);
+
+    // Shared data structures for all AutoVectorizations, to reduce allocations
+    // of large arrays.
+    VSharedData vshared;
```
Comment (vnkozlov) [inline src/hotspot/share/opto/loopnode.cpp:4872]: So it is local for each `build_and_optimize()` call and space will be freed by destructor.

File: src/hotspot/share/opto/vectorization.hpp [hunk 23]
```diff
@@ -48,15 +48,128 @@ class VTrace : public StackObj {
...
+public:
+  VSharedData() :
+    _arena(mtCompiler),
+    _node_idx_to_loop_body_idx(&_arena, estimated_node_count(), 0, 0)
+  {
+  }
+
+  GrowableArray<int>& node_idx_to_loop_body_idx() {
+    // Since this is a shared resource, we clear before every individual use.
+    _node_idx_to_loop_body_idx.clear();
```
Comment (vnkozlov) [inline src/hotspot/share/opto/vectorization.hpp:156]: I think it should be explicit `VSharedData::clear()` method called in `auto_vectorize()`. Otherwise much later someone will have hard time to find place where the space is cleared.

**Review by vnkozlov — COMMENTED** [review]
Few comments

File: src/hotspot/share/opto/loopnode.cpp (hunk 22 above)
Comment (eme64) [reply src/hotspot/share/opto/loopnode.cpp:4872]: Yes, exactly.
Before my change, we had a SuperWord object per `build_and_optimize()` that allocated all the data structures.
So the scope is now still the same.
Except that before it all went to the `comp_arena`.

So before, we might have allocated those data structures multiple times (once per `build_and_optimize`), and grown `comp_arena` each time. Now we put it to a dedicated arena, which is freed in the destructor. So the memory usage should be a little lower that way.

File: src/hotspot/share/opto/vectorization.hpp (hunk 23 above)
Comment (eme64) [reply src/hotspot/share/opto/vectorization.hpp]: @vnkozlov done.

**Review by vnkozlov — APPROVED** [review]
This looks good.

**Comment by eme64** [comment]
Thanks @vnkozlov @rwestrel for the review!
/integrate
